bundles/io/org.eclipse.smarthome.io.net.security/src/main/java/org/eclipse/smarthome/io/net/security/internal/ServiceConfiguration.java

```diff
@@ -0,0 +1,265 @@
+    String domains = "";
+    String psks = "";
+    String keystorePasswords = "";
+    String privateKeyPasswords = "";
+
+    // Self-signed certificate configuration data
+    public String organizationalUnit = "Eclipse Smarthome";
+    public String organization = "Eclipse Foundation, Inc.";
+    public String city = "Ottawa";
+    public String state = "Ontario";
+    public String country = Locale.getDefault().getCountry();
+    public BigDecimal validityInDays = BigDecimal.valueOf(365 * 3);
+
+    // Algorithm specific for new RSA or EC public/key pair
+    public String algorithmTLS = "RSA";
+    public String algorithmDTLS = "EC";
+    public String algorithmSigningTLS = "SHA1WithRSA";
+    public String algorithmSigningDTLS = "SHA256withECDSA";
+    public BigDecimal keysizeTLS = BigDecimal.valueOf(2048);
+    public BigDecimal keysizeDTLS = BigDecimal.valueOf(256);
+
+    // Management configuration
+    public Boolean autoRefreshInvalidCertificate = false;
+
+    // Parsed configuration contexts
+    protected Map<String, ContextConfiguration> contextConfigurations = new HashMap<>();
+    private File baseDir;
+
+    /**
+     * Parse the configuration fields, that contain multiple context-depending values and create
+     * corresponding {@link ContextConfiguration} objects.
+     *
+     * @param messageDigest A message digest for creating hash values over keystore files.
+     * @param baseDir The ESH home directory, used as base directory for determining keystore file paths.
+     */
+    public void parseContexts(MessageDigest messageDigest, File baseDir) {
+        this.baseDir = baseDir;
+        contextConfigurations.clear();
+
+        final Charset charSet = StandardCharsets.ISO_8859_1;
+        final Decoder decoder = Base64.getDecoder();
+        final Class<ContextConfiguration> c = ContextConfiguration.class;
+
+        try {
+            parseForField(c.getDeclaredField("domain"), domains, null);
+            parseForField(c.getDeclaredField("keystorePassword"), keystorePasswords,
+                    orig -> charSet.decode(ByteBuffer.wrap(decoder.decode(orig))).array());
+            parseForField(c.getDeclaredField("privateKeyPassword"), privateKeyPasswords,
+                    orig -> charSet.decode(ByteBuffer.wrap(decoder.decode(orig))).array());
+            parseForField(c.getDeclaredField("psk"), psks, orig -> decoder.decode(orig));
+        } catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        for (ContextConfiguration context : contextConfigurations.values()) {
+            context.keyStoreUpdated(messageDigest);
+        }
+    }
+
+    @FunctionalInterface
+    public interface CheckedFunction<T, R> {
+        R apply(T t) throws IllegalArgumentException;
+    }
+
+    protected void parseForField(Field field, String data, CheckedFunction<String, ?> mapValue)
```

Comment:
Imho this method is very hard to read and understand. With my proposal of a configuration with kay/value pairs separate for every context this would be a straight forward implementation of creating an filling the ContextConfiguration.

Reply:
I agree, it is not easy to read